### ISSUE TITLE
Slow BatchNormalization layers (CPU tested)

### ISSUE THREAD

**Issue by Zebreu** [fchollet/keras#1309]
Hi,

I added BatchNormalization layers to my model and it suddenly took much more time to train. 

It takes 561 seconds for one epoch with them:
Epoch 1/1
4096/4096 [==============================] - 561s - loss: 0.0946 - acc: 0.9006

This is if I comment out the BatchNorm layers (186 seconds):
Epoch 1/1
4096/4096 [==============================] - 186s - loss: 4.5043 - acc: 0.5933

I wrote on the user group and someone informed me that adding Dropout slows it down further, but even without Dropout this is much slower (above 500 seconds too).

Is it a CPU-related issue? Or is such a slowdown expected? Below is my model.

```
    model = Sequential()
    model.add(Convolution2D(32, 11, 11, subsample=(4,4),input_shape=(3,227,227)))
    model.add(PReLU())
    model.add(BatchNormalization())

    model.add(Convolution2D(64, 5, 5, subsample=(2,2)))
    model.add(PReLU())
    model.add(BatchNormalization())

    model.add(Convolution2D(64, 3, 3))
    model.add(PReLU())
    #model.add(BatchNormalization())
    model.add(MaxPooling2D((2,2), strides=(2,2)))

    model.add(Flatten())
    model.add(BatchNormalization())
    model.add(Dense(400))
    model.add(PReLU())
    model.add(BatchNormalization())
    model.add(Dropout(0.25))

    model.add(Dense(400))
    model.add(PReLU())
    model.add(BatchNormalization())
    model.add(Dropout(0.25))

    model.add(Dense(3, activation='softmax'))
    optimizer = keras.optimizers.RMSprop(lr=0.0005, rho=0.9, epsilon=1e-6)

    model.compile(loss='categorical_crossentropy',optimizer=optimizer)

```

**Comment by keunwoochoi** [comment]
Not sure if it's relevant but should BN be applied before activations?

**Comment by kmul00** [comment]
Hi,

Any update on this ?

The `BatchNorm` is Keras is till tremendously slow. Is there any way to alleviate the problem ?

And it increases linearly with batch size. So even if my GPU is under utilized, I cannot increase my batch size, since that would make it even more slower.

**Comment by iaroslav-ai** [comment]
+1 I observe same thing

**Comment by nouiz** [comment]
In Theano, we juste merged a wrapper of cuDNN batch norm that must be used
manually:

http://deeplearning.net/software/theano_versions/dev/library/sandbox/cuda/dnn.html#batch-normalization

It speed up the compilation and execution time.

So if you or someone else modify keras to use it, it would be great.

**Comment by EderSantana** [comment]
Nice @nouiz !!!

@fchollet @farizrahman4u tensorflow also has batch norm by default, should we benchmark this and create a batch_norm backend?

**Comment by farizrahman4u** [comment]
If its faster, then we defenitely should.

**Comment by wanderer2** [comment]
I just observed almost a 10x slowdown when using batch normalization using GPU. I am using the Theano backedn, CUDA 8 RC and CuDNN.

**Comment by mdering** [comment]
Has anyone made any progress on this? I'm working on something for myself but I'm hesitant to try to merge it because of style issues and the like. Has anyone started this on their own?

**Comment by mdering** [comment]
Has anyone attempted this? I would like to use it in my implementation (though it looks like the theano cudnn BN doesnt support 5 dimensional data yet)

**Comment by dk1013** [comment]
Have encountered the same problem with Theano backend + GPU (980 Ti). At least two times slower than models without BatchNormalization.

**Comment by nouiz** [comment]
Batch norm have a computation cost. Be sure to use cudnn 5.1. In particular
for Pascal GPU.

I don't know if Keras finished to wrap Theano cudnn batch norm or not. Make
sure that it use it.

**Comment by farizrahman4u** [comment]
Question : Is batchnorm actually worth all this trouble?

**Comment by EderSantana** [comment]
@farizrahman4u I really wish it was not... but at least for GANs, it is mandatory.

**Comment by fchollet** [comment]
Yes Keras wraps cuDNN batch norm when available on the Theano side.

You could replace BN with gradient normalization or activation
normalization.

**Comment by dk1013** [comment]
@nouiz Have updated my cudnn from v5005 to v5103, but observed no performance gain.
Is cudnn important for BatchNormalization?
I also tested the same model with tensorflow backend (after adjusting image_dim_ordering conventions etc) and found tensorflow runs way faster than theano.

**Comment by jiqiujia** [comment]
I wonder if there is any progress on this issue?

**Comment by nouiz** [comment]
Yes. Update Theano to the dev version and a "recent" enought keras will use
the new Theano interface that should have this fast from Theano side. So it
work on the CPU and the GPU with and without CuDNN. If CuDNN is there, it
will be faster.

I'm not sure which recent enough mean for keras. For Theano the master of
Theano 0.9beta1 is recent enough.
